### DISCUSSION TITLE
Setting aliases within a closure (sourcing from Bash)

### DISCUSSION THREAD

**Discussion by tgross35** [nushell/nushell#12842] [Q&A]
Similar to the pattern in https://github.com/nushell/nushell/issues/4103, I am trying to let bash parse my aliases file then load it into Nushell. This is what I have:

```nu
bash -c $"source ($env.HOME)/.shared-config/profile-common.sh && alias" |
  lines |
  parse "alias {k}={v}" |
  filter { |x| ($x.k not-in ['..', 'gcr']) } |
  each { |x| alias $x.k = do $x.v }
```

However, `$x.k` flags with "﻿﻿alias name can't be a number, a filesize, or contain a hash # or caret ^" (none of these apply). Changing `$x.k` to a fixed string for testing completes successfully, but does not set the alias.

Is there a better way to do this?

**Comment by ghost** [comment]
This is one of the weaknesses of nushell for which i have not found a perfect workaround (but maybe someone else has):

you cant export aliases, etc in closures due to the scope:
```
$ def test [] { alias foo = echo foo }
$ test
$ foo
Error: nu::shell::external_command

  × External command failed
   ╭─[entry #1:1:1]
 1 │ foo
   · ─┬─
   ·  ╰── did you mean 'for'?
   ╰────
  help: No such file or directory (os error 2)
```
(not just functions, but also `| each {}` blocks, etc)

a way to circumvent the scope problem is by writing it into a script and then using `use`.
but it does not matter where in a script use is called the imported file has to be created before the script is executed.

the workaround within the context of a config file for this is by creating a temporary script file in `env.nu` and then importing it in `config.nu`.

`env.nu`:
```nu
open -r ~/.bashrc
| lines
| parse -r "^[ \t]*alias (?P<from>[^=]+)=(?P<to>.*)$"
| each {|i| $"alias ($i.from) = do ($i.to)"}
| prepend "export-env {"
| append "}"
| str join "\n"
| save -rf /tmp/bash_alias_export_to.nu
```
`config.nu`:
```nu
use /tmp/bash_alias_export_to.nu
```

and the only workaround im aware of for other scripts is by generating a script containing everything the script intends to do inside of a string and then executing the string in its own context using `nu -c`.

```nu
nu -c (
  open -r ~/.bashrc
  | lines
  | parse -r "^[ \t]*alias (?P<from>[^=]+)=(?P<to>.*)$"
  | each {|i| $"alias ($i.from) = do ($i.to)"}
  | append "some other command"
  | str join "\n"
)
```

altough id recommend using custom aliases for nushell (at least for the most part).

or what i do with other programs: maintain a `csv` (or different) parsable shared config file, which allows defining exceptions and then compile it into the different config file formats (using nushell as compiler).  
this for example is the neovim part of my neovim/helix keybind config:
```nu
open ./keybinds.csv
| where nvim == "true"
| each {|i|
  $"map\(\"($i.mode)\", \"($i.key)\", \"($i.action)\");"
}
| prepend "#include \"./nvim.h\"\nint luaopen_config_keybinds() {"
| append 'return 1;}'
| str join "\n"
| save -r src/keybinds.c

^gcc -o lua/config/keybinds.so -std=gnu11 -O2 src/keybinds.c -shared -fPIC
rm src/keybinds.c
```
